### PR TITLE
Faster implementation of Float64(::BigInt), Float32(::BigInt) and Float16(::BigInt)

### LINKED ISSUE #31293: BigInt to Float conversions are unnecessarily slow
```
julia> using BenchmarkTools

julia> x = big"10"^300;

julia> @btime Float64($x)
  468.070 ns (6 allocations: 328 bytes)
1.0e300

julia> @btime Float64(BigFloat($x, precision=precision(BigFloat)))
  74.334 ns (2 allocations: 112 bytes)
1.0e300
```

This could probably be done even more efficiently by doing bit-shifting on the limbs.

### PR DESCRIPTION
**Pull request by narendrakpatel**
Closes #31293 
Implements a faster version of `Float64(::BigInt)`, `Float32(::BigInt)` and `Float16(::BigInt)`.
Rounding Behaviour: RoundNearest

### CODE REVIEW

File: base/gmp.jl [hunk 1]
```diff
@@ -396,7 +396,36 @@ function (::Type{T})(n::BigInt, ::RoundingMode{:Nearest}) where T<:CdoubleMax
     x
 end
 
-Float64(n::BigInt) = Float64(n, RoundNearest)
+function Float64(x::BigInt)
+    x == 0 && return 0.0
+    if Int == Int64
```
Comment (simonbyrne) [inline base/gmp.jl]:
```suggestion
    if Base.GMP.Limb == UInt64
```

**Comment by simonbyrne** [comment]
In general looks pretty good: would be nice if we could make this more generic, as we will need to repeat it for Float32 and Float16.

Need to also check that we cover good test cases as well.

**Comment by narendrakpatel** [comment]
Currently, I am trying to refactor the code to remove unnecessary computations and the code 'prettier'.

For generalization purposes, I am attempting to write code by using a variable `prec` which will assume the value 24 in the case of Float32 and 11 in the case of Float16. I believe since a general method can be made for Float32 and Float16 (but maybe not for Float64 as it can consume 3 Limbs).
Any suggestions? 

Also work related to Float32 and Float16 should be done in this issue itself or a separate issue will be opened?

**Comment by simonbyrne** [comment]
Lets keep it all in the same pull request.

To help make it generic, there are various utility functions that can be of use, e.g.:
https://github.com/JuliaLang/julia/blob/91151ab871c7e7d6689d1cfa793c12062d37d6b6/base/float.jl#L571-L582
https://github.com/JuliaLang/julia/blob/91151ab871c7e7d6689d1cfa793c12062d37d6b6/base/float.jl#L846-L873

**Comment by narendrakpatel** [comment]
I will make sure to use them as much as possible.

**Comment by narendrakpatel** [comment]
@simonbyrne I have implemented the conversion methods with `RoundNearest` rounding behavior.
I had a doubt regarding tests. Currently, there is no `test/gmp.jl` file so shall I create a new file and write tests in it or shall I write tests in `float.jl`?

**Comment by rfourquet** [comment]
>  there is no test/gmp.jl

But there is "test/bigint.jl".

**Comment by narendrakpatel** [comment]
Yes, there is. :sweat_smile: 
I will just append tests in it. Thanks, @rfourquet

**Comment by narendrakpatel** [comment]
@simonbyrne is this okay?

**Comment by simonbyrne** [comment]
In general, it looks pretty good. For each type`T` it would be a good idea to check:
```
n = exponent(floatmax(T))
@test T(big"2"^(n+1)) === T(Inf)
@test T(big"2"^(n+1) - big"2"^(n-precision(T))) === T(Inf)
@test T(big"2"^(n+1) - big"2"^(n-precision(T))-1) === floatmax(T)
```

**Comment by narendrakpatel** [comment]
@simonbyrne I have updated the tests. Is there any way I can generalize other tests too? The idea I used is to test for conversion of `BigInts` which use a different number of `Limb`.
If there is no issue in the current format of tests then this PR might be ready to merge.

**Comment by simonbyrne** [comment]
> @simonbyrne I have updated the tests. Is there any way I can generalize other tests too? The idea I used is to test for conversion of `BigInts` which use a different number of `Limb`.
> If there is no issue in the current format of tests then this PR might be ready to merge.

File: base/gmp.jl [hunk 2]
```diff
@@ -396,9 +396,73 @@ function (::Type{T})(n::BigInt, ::RoundingMode{:Nearest}) where T<:CdoubleMax
...
+            y += n > (precision(Float64) - 32) ? 0 : (unsafe_load(x.d, x.size-2) >> (10+n))
+        end
+        y = (y + 1) >> 1
+        y &= ~UInt64(trailing_zeros(x) == (n-54 + (xsize-1)*BITS_PER_LIMB))
+        d = ((n+1021) % UInt64) << 52
+        z = reinterpret(Float64, d+y)
+        z = ldexp(z, (xsize-1)*BITS_PER_LIMB)
+    end
+    signbit(x.size) && return -z
+    return z
```
Comment (simonbyrne) [inline base/gmp.jl]: Change the last two lines to
```
return signbit(x.size) ? -z : z
```

File: base/gmp.jl [hunk 3]
```diff
@@ -396,9 +396,73 @@ function (::Type{T})(n::BigInt, ::RoundingMode{:Nearest}) where T<:CdoubleMax
     x
 end
```
Comment (simonbyrne) [inline base/gmp.jl]: delete this function

File: base/gmp.jl [hunk 4]
```diff
@@ -396,9 +396,73 @@ function (::Type{T})(n::BigInt, ::RoundingMode{:Nearest}) where T<:CdoubleMax
...
 
-Float64(n::BigInt) = Float64(n, RoundNearest)
-Float32(n::BigInt) = Float32(n, RoundNearest)
-Float16(n::BigInt) = Float16(n, RoundNearest)
```
Comment (simonbyrne) [inline base/gmp.jl]: add these back in

File: base/gmp.jl [hunk 5]
```diff
@@ -396,9 +396,73 @@ function (::Type{T})(n::BigInt, ::RoundingMode{:Nearest}) where T<:CdoubleMax
...
+function Float64(x::BigInt)
```
Comment (simonbyrne) [inline base/gmp.jl]:
```suggestion
function Float64(x::BigInt, ::RoundingMode{:Nearest})
```
and similar for `Float32`/`Float16`.

**Review by simonbyrne — CHANGES_REQUESTED** [review]
We can remove the existing definition of round-to-nearest conversion, and make it this one.

After this, I think it's good to go.

**Comment by narendrakpatel** [comment]
@simonbyrne updated the PR.

**Review by simonbyrne — APPROVED** [review]

**Comment by narendrakpatel** [comment]
If this is okay, can we merge this? (before the 1.2 freeze :smile: )

File: base/gmp.jl [hunk 6]
```diff
@@ -374,26 +374,69 @@ function (::Type{T})(n::BigInt, ::RoundingMode{:Up}) where T<:CdoubleMax
...
+            y += (unsafe_load(x.d, xsize-1) % UInt64) >> (n-22)
+            y += n > (precision(Float64) - 32) ? 0 : (unsafe_load(x.d, x.size-2) >> (10+n))
         end
+        y = (y + 1) >> 1
+        y &= ~UInt64(trailing_zeros(x) == (n-54 + (xsize-1)*BITS_PER_LIMB))
+        d = ((n+1021) % UInt64) << 52
+        z = reinterpret(Float64, d+y)
+        z = ldexp(z, (xsize-1)*BITS_PER_LIMB)
+    end
+    return signbit(x.size) ? -z : z
```
Comment (rfourquet) [inline base/gmp.jl]: this is `flipsign(x.size, z)`

File: base/gmp.jl [hunk 7]
```diff
@@ -374,26 +374,69 @@ function (::Type{T})(n::BigInt, ::RoundingMode{:Up}) where T<:CdoubleMax
...
+    if xsize*BITS_PER_LIMB > 1024
+        z = Inf64
+    elseif xsize == 1
+        z = Float64(unsafe_load(x.d))
+    elseif Limb == UInt32 && xsize == 2
+        z = Float64((unsafe_load(x.d, 2) % UInt64) << BITS_PER_LIMB + unsafe_load(x.d))
+    else
+        y1 = unsafe_load(x.d, xsize) % UInt64
+        n = 64 - leading_zeros(y1)
+        y = y1 >> (n-54)
```
Comment (rfourquet) [inline base/gmp.jl]: Although I have played myself bit-fiddling with floats, I had to think twice to understand what is this `54`. Is it `precision(Float64)+1` ? At least a comment could be helpful, to prepare the reader to interpret this and upcoming magic numbers.

File: base/gmp.jl [hunk 8]
```diff
@@ -374,26 +374,69 @@ function (::Type{T})(n::BigInt, ::RoundingMode{:Up}) where T<:CdoubleMax
...
-            if r <= -h
-                return prevfloat(x)
-            elseif r >= h
-                return nextfloat(x)
-            end
+            y += (unsafe_load(x.d, xsize-1) % UInt64) >> (n-22)
+            y += n > (precision(Float64) - 32) ? 0 : (unsafe_load(x.d, x.size-2) >> (10+n))
         end
+        y = (y + 1) >> 1
+        y &= ~UInt64(trailing_zeros(x) == (n-54 + (xsize-1)*BITS_PER_LIMB))
```
Comment (rfourquet) [inline base/gmp.jl]: It's not so trivial to understand the algorithm used here... Would you have a comment or two to help parse these operations?

Comment (narendrakpatel) [reply base/gmp.jl]: Can you please check the latest commits and give feedback.

Comment (rfourquet) [reply base/gmp.jl]: That's very helpful, thank you!

**Comment by simonbyrne** [comment]
Looks like there is a test failure on 32-bit Linux:
```
Test Failed at /buildworker/worker/tester_linux32/build/share/julia/test/bigint.jl:447
  Expression: Float64(-x) == -(Float64(x))
   Evaluated: -1.1579208923884626e77 == -1.1579208923731622e77
```

**Comment by narendrakpatel** [comment]
@simonbyrne All the tests are now passing except the following:
* *package_freebsd64* : error during make
* *appveyor* : error during running test `Pkg/pkg`
* *travis-ci* : failed due to connection time out.

**Comment by narendrakpatel** [comment]
@simonbyrne If this is okay, can this be merged?

**Pull request merged**